### PR TITLE
feat: add buildTarget/format method and canFormat capability

### PR DESCRIPTION
**Pull request by danielleontiev**
Following the discussion here https://github.com/scalameta/metals-feature-requests/issues/334 this PR extends BSP to support formatting of source files.

I am not sure whether I have correctly approached unit testing in the project. Please pay attention to it because mostly, I have done tests by "mocking" the similar `builtTarget/cleanCache` request by copy-pasting its test code and changing the method. Also, please let me know if I should fix comments by adding additional commits or if it is allowed to force-push.

Motivation

- Some build servers that implement BSP can be aware of source code formatting but there is no way to trigger formatting in abstract way in build-tool-transparent way. Example: `scala-cli` is universal tool for running/compiling/testing (and formatting as well) Scala code. It also implements BSP. `scala-cli` does not require to have `.scalafmt.conf` file to format code because it has built-in default configuration. It is possible to format source files with `scala-cli` by running `scala-cli fmt` but it should be done via CLI interface from the shell. If one works with Scala code using `scala-cli` as BSP server and tries to trigger "code formatting" from LSP client/IDE it will require some additional movements - for example, if using Metals it's necessary to create `.scalafmt.conf` file to use formatting capability of Metals (because Metals does not know that `scala-cli` knows how to do code format and that it does not require `.scalafmt.conf`). We also can imagine the case where `scala-cli` or similar tool even do not use `scalafmt` under the hood - for example, IntelliJ has built-in code formatter that is not `scalafmt`.
- Build server can manage source files written in different languages. Example: `sbt` can build projects based on both Scala and Java sources. If you work with Scala as a primary language in the project the only way to ensure that Java sources are formatted properly is to use separate tool or have IDE with Java support (IntelliJ) or install Java LSP client for Java if you use VSCode/Nvim/etc. Since `sbt` is already aware of Java sources in the project it seems to be a nice fit for it to implement Java source formatting capability - not in the `sbt` itself but for example using `sbt` plugin. Another example is `ScalaPB`. If you use it in `sbt` it comes in form of plugin that runs Protobuf compiler for Scala from `.proto` sources in your project. `ScalaPB` sbt plugin also seems a very good place to implement `.proto` formatting while currently it is possible only using external tools or IDE/LSP client.

### CODE REVIEW

File: bsp4j/src/main/java/ch/epfl/scala/bsp4j/Protocol.xtend
Comment (adpi2) [reply bsp4j/src/main/java/ch/epfl/scala/bsp4j/Protocol.xtend:743]: Is scala-cli able to format a subset of files already?

**Review by adpi2 — COMMENTED** [review]
Thanks for this PR. I agree that we need the ability to format individual files. Also I wonder if the `FormatResult` structure is really useful.

File: bsp4j/src/main/java/ch/epfl/scala/bsp4j/Protocol.xtend
Comment (danielleontiev) [reply bsp4j/src/main/java/ch/epfl/scala/bsp4j/Protocol.xtend:743]: image

Yes it does - at least using CLI. In the screenshot above I've done the formatting of only one file in the directory


Another question is how much preciese show this option be. We can accept a list of files for each target for example - in that case the parameters would look like this:

```
[
  {
    "targetId": ...,
    "filesToFormat": [ // list of files and/or directories ]
  }
]
```

If list `filesToFormat` is empty - then the whole target should be formatted.

Another option is the following


```
[
  {
    "targetId": ...,
    "fileToFormat": "individual file (optional)"
  }
]
```

In this example we have only one optional parameter - if it passed then the passed file should be formatted, otherwise the whole target should be formatted. It looks more simple and it's useful in a particular case - to format the file you are currently working on in the editor/IDE

Comment (adpi2) [reply bsp4j/src/main/java/ch/epfl/scala/bsp4j/Protocol.xtend:743]: I would choose the first option with an array of files because that's more flexible and probably not harder to implement.

**Review by adpi2 — APPROVED** [review]
Sounds good to me!

@jastice Do you want to review before we merge?

**Comment by danielleontiev** [comment]
> > Sounds good to me!
> > @jastice Do you want to review before we merge?
> 
> I want to review, will do that later this week

Hi @agluszak! Have you got a chance to take a look at the PR?

**Comment by agluszak** [comment]
Yes, but I need to discuss it further with my team because of #470. I suggest that we don't hurry with merging this one.

**Comment by adpi2** [comment]
@agluszak This PR is almost 4 weeks old already. What is it exactly that you think should be discussed? In the current version of BSP the notion of build target is the central notion. So I think it is fine to merge this PR as it is. We can change it later, in BSP 3 if need be.

**Comment by jastice** [comment]
Sorry to keep you waiting so long. From our perspective, while we wouldn't want to outright block this change, we'd like to add these points for consideration:

- does this feature need to be in the base protocol at this point? As I understand, the endpoint is currently only supported by scala-cli. Do other build tools have an intention of supporting it? Otherwise, maybe it's better to just implement it for the specific server and client that have a need.
- adding this to the base protocol does incur a bit of a maintenance burden to existing implementations for updating the code for the new implementations

From a technical perspective: is a build target the best unit of formatting? the change does now support formatting individual files, but does the target need to be part of the request at all? It could be something like LSPs [`textDocument/formatting`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_formatting) instead that operates on document level (but possibly operating in bulk). Additionally, files may belong to multiple build targets. On the other hand, it seems reasonable to have a way to format file sin bulk without listing them all.

If you feel the benefits of the change outweigh these points, I'm okay with merging it. Alternatively, I propose adding it to a separate section as extension to document its existence until there are more tools ready to implement it.

**Comment by danielleontiev** [comment]
> does this feature need to be in the base protocol at this point? As I understand, the endpoint is currently only supported by scala-cli. Do other build tools have an intention of supporting it? Otherwise, maybe it's better to just implement it for the specific server and client that have a need.

I am OK with making it as protocol extension - besides that I can image the adoption of this endpoint not only in `scala-cli` (for example, `sbt`'s plugin for `scalafmt` or `protobuf` also can benefit from adopting it), but of course it is arguably not something that fits to every BSP implementation - so, adding it as extension makes sense.

> As I understand, the endpoint is currently only supported by scala-cli

It's not supported by `scala-cli` yet but if the change will be accepted to BSP protocol I will try to push the changes to `scala-cli`'s BSP implementation since the use case of `scala-cli` was the reason of creating this PR in the first place.

> is a build target the best unit of formatting?

I do not know to be honest - I do not have a strong arguments for/against making `buildTarget` as unit of formatting. But I think that it's important to support two use cases: (1) formatting individual files and (2) formatting "everything". If we use `buildTarget/format` without parameters (currently proposed way of formatting "everything") then BSP implementation should format all sources inside build target. I also can imagine the achieving the same using `textDocument/format` endpoint - it may accept a list of `TextDocumentIdentifier` as a request which is defined like this:

```scala
final case class TextDocumentIdentifier(
    uri: Uri
)
```

The `uri` inside could be individual file, supporting use case (1), as well as `BuildTarget.baseDirectory`, `BuildTargetIdentifier.uri`, list of `SourceItem.uri` obtainted from `buildTarget/sources` or even `InitializeBuildParams.rootUri` - so, it seems that there are a plenty of ways of covering use case (2) even without making `buildTarget` as a unit of formatting.

Also taking the #470 into account I agree that having `textDocument/format` endpoint may be easier to evolve in the future.

To sum up, I am ready to adjust the PR to create `textDocument/format` taking `[]TextDocumentIdentifier` as parameter method instead of initially proposed `buildTarget/format`. Also it should be changed to be a not a part of main protocol but extension of the main protocol.

Please, comment how do you feel about such adjustment - if it's OK then I will change the PR

**Comment by danielleontiev** [comment]
Another question that arises if we do not use `buildTarget` is how to show the capability of formatting to the BSP client?

I see that `BuildServerCapabilities` have parameters like `jvmTestEnvironmentProvider` therefore the object can be extended with `canFormat` field. On the other hand it makes the main protocol aware of the existence of "formatting" extension which may be an argument against such change.  `jvmTestEnvironmentProvider` also makes the main protocol aware of `buildTarget/jvmTestEnvironment` extension - but I do not know whether it will be OK to extend the object with `canFormat` or it was an exceptional case for `buildTarget/jvmTestEnvironment` extension.

**Comment by adpi2** [comment]
> does this feature need to be in the base protocol at this point? As I understand, the endpoint is currently only supported by scala-cli. Do other build tools have an intention of supporting it? Otherwise, maybe it's better to just implement it for the specific server and client that have a need.

In this PR, formatting is implemented as a capability so that the build tool can decide to implement it or not. See, in `BuildServerCapabilities` there is a `formatProvider` that can be left empty if the server does not support this capability, and, in `buildTarget`, there is a `canFormat` capability that is `false` by default.

> adding this to the base protocol does incur a bit of a maintenance burden to existing implementations for updating the code for the new implementations

I don't think they need to update anything, because `formatProvider` and `canFormat` have default values. They can be omitted in the messages.

> Alternatively, I propose adding it to a separate section as extension to document its existence until there are more tools ready to implement it.

We have extensions and capabilities. On one hand, a capability is shared between many languages but may or may not be implemented in some build server or used in some build client. On the other hand, an extension is specific to a certain language, like Scala, or Rust or cpp. Formatting, not being specific to any language, should be a capability not an extension.

That being said we can maybe add a `staging` or `experimental` section in the based protocol, which would mean that a part of the spec could change in the near future could be removed or changed. But, IMO, adding something to the base protocol does not mean it is there forever and won't change. We should feel open to deprecation if need be.

> From a technical perspective: is a build target the best unit of formatting? the change does now support formatting individual files, but does the target need to be part of the request at all? It could be something like LSPs [textDocument/formatting](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_formatting) instead that operates on document level (but possibly operating in bulk). Additionally, files may belong to multiple build targets. On the other hand, it seems reasonable to have a way to format file sin bulk without listing them all.

Yes I think the build target is the best unit of formatting because:
- the build server can only format the files that belong to a build target
- the capability is enabled/disabled at the build target level

As I see things, a build target is the main unit of the BSP protocol, and so all requests that perform an operation (compile, test, run, format) should be scoped to a build target. In other words, any operation that is not scoped to a build target should probably not be part of build server protocol.

**Comment by jastice** [comment]
@danielleontiev 
> It's not supported by scala-cli yet but if the change will be accepted to BSP protocol I will try to push the changes to scala-cli's BSP implementation since the use case of scala-cli was the reason of creating this PR in the first place.

What do you think of creating an implementation before adding it to the protocol? Any client/server is free to implement compatible additions to the protocol, after all. This way we could see if the mechanism pans out as intended, and would make it easier to iterate in case it doesn't.

> The uri inside could be individual file, supporting use case (1), as well as BuildTarget.baseDirectory, BuildTargetIdentifier.uri, list of SourceItem.uri obtainted from buildTarget/sources or even InitializeBuildParams.rootUri - so, it seems that there are a plenty of ways of covering use case (2) even without making buildTarget as a unit of formatting.

This approach sounds good to me.

> Another question that arises if we do not use buildTarget is how to show the capability of formatting to the BSP client?

A good question. A build server may have the ability to format some files but not others. For example only Scala, but not JS files. This is also an argument for having formatting on `textDocument` level, even though in BSP targets are kind of associated with a specific language.

Broadly I would propose: 
- a formatting capability on build server level (rather than build target), informing which file types can be formatted
    - e.g. `canFormat: [scala, java, js]`
- a `textDocument/format` request operating on lists of files or directories
    - a response that tells you if formatting was successful or unsupported

just an idea, what do you think?


@ckipp01 
> Think large repos that can have thousands and thousands of files. I know you can send an empty array for formatting everything but more than likely that may not be wanted resulting in an array of thousands and thousands of URIs when a user might just want to leave out one target.

Not sure I understand. Are you saying that a user may want to format all targets but one? That doesn't seem like a very likely use case. I can think of these likely cases and way of handling them in above proposal:
- format specific files (send file array)
- format everything in directories (send directory array)
- format everything the server knows (send empty array in request)
- format everything in target (send array of files in target, which client generally knows)

I don't think client should be required in any of these cases to specify both target and files

@adpi2 
> I don't think they need to update anything, because formatProvider and canFormat have default values: they can be omitted in the messages.
Fair enough

> We have extensions and capabilities. On one hand, a capability is shared between many languages but may or may not be implemented in some build server or used in some build client. On the other hand, an extension is specific to a certain language, like Scala, or Rust or cpp. Formatting, not being specific to any language, should be a capability not an extension.

I don't think an extension is necessarily about a language. For instance we have Maven and sbt extensions. But aside of this, the concepts are a bit fuzzy anyway. 

> That being said we can maybe add a staging or experimental section in the base protocol, which would mean that a part of the spec could change or be removed. But, IMO, adding something to the base protocol does not mean it is there forever and won't change. We should feel open to deprecation if need be.

As long as we have a limited number of implementors, incompatibly evolving it is doable, but it still requires coordination effort. So I'm in favor of some kind of `experimental` status for new features.

> Yes I think the build target is the best unit of formatting because:
> the build server can only format the files that belong to a build target
> the capability is enabled/disabled at the build target level

I don't think these are true. The server could conceivably format anything it has access to. Formatting is not generally  dependent on the sets of dependencies, which is the main point of a target. It's unlike compile/test/run in that way.

> In other words, any operation that is not scoped to a build target should probably not be part of build server protocol.

I also don't think this necessarily holds, though I agree BSP as it is mostly works on targets.

**Comment by adpi2** [comment]
> a formatting capability on build server level (rather than build target), informing which file types can be formatted
e.g. canFormat: [scala, java, js]

This is already implemented like that in this PR.

> I don't think an extension is necessarily about a language. For instance we have Maven and sbt extensions. But aside of this, the concepts are a bit fuzzy anyway.

It does not look so fuzzy to me. The current sbt extension is more about the sbt language (how to compile build.sbt file) than about build tool itself. And the Maven extension is about the dependencies of the JVM languages.

> I don't think these are true. The server could conceivably format anything it has access to.

How do we know that the server has access to those files? We know because they belong to a build target, as source files. A scala file is the source file of a scala target, a java file is the source file of a java target, a build.sbt is the source file of an sbt target, a protobuf file is the source file of a protobuf target...

> Formatting is not generally dependent on the sets of dependencies, which is the main point of a target. It's unlike compile/test/run in that way.

Why do we want the dependencies to be the main point of a target? I would rather say that a build target is something that the build tool can build, which does often imply that there are some dependencies but not always. I can think of some template languages that have some source files that we can format/compile, but no dependencies.

I think the BSP model should be: a build (or workspace) is composed of build targets and the server can perform some operations on them: compiling, testing, running, and/or formatting.

**Comment by jastice** [comment]
> a formatting capability on build server level (rather than build target), informing which file types can be formatted
e.g. canFormat: [scala, java, js]

> This is already implemented like that in this PR.

How so? In this PR we have a `canBuild: Boolean` on target level.

> How do we know that the server has access to those files? We know because they belong to a build target, as source files. A scala file is the source file of a scala target, a java file is the source file of a java target, a build.sbt is the source file of an sbt target, a protobuf file is the source file of a protobuf target...

Not wrong, but (1) source files can belong to multiple targets, (2) BSP has the concept of a workspace - not all files in the workspace are necessarily part of a target, and the client shouldn't need to provide the target for a file to format it.

> Why do we want the dependencies to be the main point of a target? I would rather say that a build target is something that the build tool can build, which does often imply that there are some dependencies but not always

I think I formulated that badly. But conceptually, the target abstraction comprises some operations you can do on it, as you say, like build, and the dependencies necessary to be able to do those operations (libraries and other targets). Building, testing, running are things you generally do on a target level - they require the whole target and its dependencies as input. Formatting is independent of anything in the target and any dependencies, you can format each file in isolation.

**Comment by adpi2** [comment]
> How so? In this PR we have a canBuild: Boolean on target level.

There is a `formatProvider` in the `BuildServerCapabilities` which contains a list of languages that the build tool can format.

> (2) BSP has the concept of a workspace - not all files in the workspace are necessarily part of a target

In the current definition, workspace is just a list of targets. Should we change that definition so that we can have some files that don't belong to any target? Do you have an example of such file in some build tool or server?

**Comment by agluszak** [comment]
> adding this to the base protocol does incur a bit of a maintenance burden to existing implementations for updating the code for the new implementations
> I don't think they need to update anything, because formatProvider and canFormat have default values: they can be omitted in the messages.

The concept of "maintenance burden" came from our JB-internal discussion. The problem here is not necessarily that existing tools would have to update, because sure, formatting capability would be opt-in. The problem is that if we decide to make formatting a core capability of BSP and later introduce changes to it, then we will break tools which would have implemented it.

And we will certainly want to evolve that. For example, in the current implementation there is no way to pass settings (or any additional context) to the formatter. Using a configuration file is just a single way to do that, some other formatting tools use flags instead.

> if using Metals it's necessary to create .scalafmt.conf file to use formatting capability of Metals

How would a server inform the client that a capability has become available, if it doesn't depend on inherent properties of the target (like test, run, etc. do), but on some external state?

> Do you have an example of such file in some build tool or server?

In Bazel there are BUILD and WORKSPACE files. They don't belong to any Bazel target (because they are used to *define* targets). We could come up with some kind of meta-targets that would contain only Bazel-related files. You said that in sbt `build.sbt` are part of an sbt target. How does that work?

> I think the BSP model should be: a build (or workspace) is composed of build targets and the server can perform some operations on them: compiling, testing, running, and/or formatting.

Well, actually there are more operations we could possible want to [run on targets](https://github.com/bazelbuild/buildtools/blob/master/buildozer/README.md): linting or [finding unused dependencies](https://github.com/bazelbuild/buildtools/blob/master/unused_deps/README.md) for example. I think we need a more general abstraction for "running tools on targets".

**Comment by danielleontiev** [comment]
Since it's hard to find a consensus here, I will try to create both implementations first, and we'll see which one works better for the `scala-cli` case. I understand that the option that works better for `scala-cli` won't necessarily be the best one to include in the BSP protocol, but at least we will have something to compare and iterate on.

**Comment by danielleontiev** [comment]
I've done both implementations, they are not able to actually format Scala files, however they do implement all protocol changes.

- https://github.com/danielleontiev/scala-cli/pull/1
- https://github.com/danielleontiev/scala-cli/pull/2

Here are some thoughts after spending some time implementing the changes:

- Adding a method (either `buildTarget/format` or `textDocument/format`) is straightforward. However, extending models of existing methods is tedious because it essentially requires defining a new `BuildServer` interface. This interface consists of non-modified methods plus modified methods, where the extended methods require a copy-paste of the relevant models. For example, adding the `canFormat` build target capability requires changes in the entire chain of models: `BuildTargetCapabilities -> BuildTarget -> WorkspaceBuildTargetsResult`.
- `buildTarget/format`:
   1. It's not very clear how `BuildServerCapabilities/formatProvider` should work together with `BuildTargetCapabilities/canFormat`. The `formatProvider` returns language IDs that it is able to format (e.g., `languageIds = [ "scala" ]`), and the build target also has a `languageIds` array. I'm unsure if it's possible for a build target to have `languageIds = [ "scala" ]` and `canFormat = false` at the same time.
   2. Another question is, for example, if a build target has `languageIds = [ "scala", "java", "proto" ]`, should `scala-cli`'s BSP return `canFormat = true` or `false`? From my perspective, it is much clearer to follow the approach of `textDocument/format`. If the BSP returns a format provider for a specific language, the client should assume that the server is able to format sources of that language.
   3. For `scala-cli`, the above questions are not very critical because it currently operates on two build targets: **main** and **test**. Therefore, in my draft implementation, I always return `canFormat = true` for both targets.
- `textDocument/format`:
  1. In terms of implementation, this was a simpler version of the above because it only required adjusting one method, `build/initialize`, to return a format provider.

At this point, I believe it would be reasonable to create an issue in the `scala-cli` project and ask the maintainers for feedback on both implementations. They may have reasons to prefer one implementation over the other. I will proceed with creating the issue.

**Comment by danielleontiev** [comment]
Sorry for keeping this for too long but after a discussion in https://github.com/VirtusLab/scala-cli/issues/2134 it was finally decided to not take either of `textDocument/format` or `buildTarget/format` approaches in `scala-cli`. Since there is no BSP implementation that will adopt BSP formatting in near future, I believe it's better to close this one and not make change ahead. Thank you all who took part in this!

**Pull request closed without merging**
